File: partition.py
# // Time Complexity :O(n)
# // Space Complexity :O(1)
# // Did this code successfully run on Leetcode :Yes
# // Any problem you faced while coding this :No

# we keep tarck of last index of each alphabet and create a window with the end as the end we caluclated - now we stop the window when the end has been reached -
class Solution(object):
    def partitionLabels(self, s):
        """
        :type s: str
        :rtype: List[int]
        """
        # O(n), O(1)
        res=[]
        hmap={}
        for i in range(len(s)):
            c=s[i]
            hmap[c]=i
        start=0
        end=0
        for i in range(len(s)):
            c=s[i]
            end=max(end, hmap[c])
            if(i==end):
                
                res.append(end-start+1)
                start=i+1
        return res



        

File: Partition.cpp
/*
Intuition: BF: We can get all partitions and check if all characters are in the same partition.

We can optimize this by trackign the ending index of all the characters. We keep expanding the partition based on incoming characters last index.

Time Complexity: O(N), N = Size of array
Space Complexity: O(N)
*/


class Solution {
public:
    vector<int> partitionLabels(string s) {
        map<char, int> map;
        for ( int i = 0; i < s.size(); i++){
            char letter = s[i];
            map[letter]= i;
        }
        
        vector<int> result;
        int start = 0;
        int end = 0;
        for (int i = 0; i < s.size(); i++){
        
            end = max( end, map[s[i]] );
            if ( i == end){
                result.push_back(end - start + 1);
                start = end +1;
            }

        }
        return result;
    }
};